src/wcl/doc.h

```diff
@@ -0,0 +1,253 @@
+
+namespace wcl {
+
+// Internal state for a doc object. It is intentionally not exposed in the
+// public API and should not be used directly.
+struct doc_state {
+  size_t newline_count = 0;
+  size_t first_width = 0;
+  size_t last_width = 0;
+
+  // Returns the merged state of this and other.
+  doc_state merged(doc_state other) {
+    doc_state merged;
+
+    merged.newline_count = newline_count + other.newline_count;
+
+    merged.last_width = other.last_width;
+    // If other doesn't have any newlines, then the left side needs to be accounted for
+    if (other.newline_count == 0) {
+      merged.last_width += last_width;
+    }
+
+    merged.first_width = first_width;
+    // If the left side doesn't have any newlines, then other side needs to be accounted for
+    if (newline_count == 0) {
+      merged.first_width += other.first_width;
+    }
+
+    return merged;
+  }
+
+  static doc_state from_string(const std::string& str) {
+    doc_state state;
+    for (auto& c : str) {
+      if (c == '\n') {
+        state.last_width = 0;
+        state.newline_count++;
+        continue;
+      }
+
+      if (state.newline_count == 0) {
+        state.first_width++;
+      }
+
+      state.last_width++;
+    }
+    return state;
+  }
+};
+
+class doc_impl_base {
+ protected:
+  // On construction this has to be set
+  size_t size_;
+
+  doc_state state_;
+
+  doc_impl_base(size_t size) : size_(size) {}
+
+ public:
+  virtual ~doc_impl_base() = default;
+
+  // methods that all RopeImpl share
+  virtual void write(std::ostream&) const = 0;
+  size_t size() const { return size_; }
+  size_t first_width() const { return state_.first_width; }
+  size_t last_width() const { return state_.last_width; }
+  size_t newline_count() const { return state_.newline_count; }
+  bool has_newline() const { return state_.newline_count > 0; }
+  doc_state state() const { return state_; }
+};
+
+class doc_impl_string : public doc_impl_base {
+ private:
+  std::string str;
+
+ public:
+  explicit doc_impl_string(std::string str) : doc_impl_base(str.size()), str(str) {
+    state_ = doc_state::from_string(str);
+  }
+  void write(std::ostream& ostream) const override { ostream << str; }
+};
+
+class doc_impl_pair : public doc_impl_base {
+  // this concats two docs together
+ private:
+  std::pair<std::shared_ptr<doc_impl_base>, std::shared_ptr<doc_impl_base>> pair;
+
+ public:
+  doc_impl_pair(std::shared_ptr<doc_impl_base> left, std::shared_ptr<doc_impl_base> right)
+      : doc_impl_base(left->size() + right->size()),
+        pair(std::make_pair(std::move(left), std::move(right))) {
+    state_ = pair.first->state().merged(pair.second->state());
+  }
+
+  void write(std::ostream& ostream) const override {
+    pair.first->write(ostream);
+    pair.second->write(ostream);
+  }
+};
+
+// Forward declare doc_builder so it can be named as a friend of doc
+class doc_builder;
+
+// `doc` is a very efficient data structure for editing strings. It is ideal to use when
```

Comment:
I realize this is an old comment but I just re-read it and it isn't true, its an efficient data structure for *constructing* strings. It's actually immutable.